client/app/visualizations/cohort/Cornelius.jsx

```diff
@@ -0,0 +1,209 @@
+/*!
+ * React port of Cornelius library (based on v0.1 released under the MIT license)
+ * Original library: http://restorando.github.io/cornelius
+ */
+
+import { isNil, isFinite, map, extend, min, max } from 'lodash';
+import moment from 'moment';
+import chroma from 'chroma-js';
+import React, { useMemo } from 'react';
+import PropTypes from 'prop-types';
+import Tooltip from 'antd/lib/tooltip';
+import { createNumberFormatter, formatSimpleTemplate } from '@/lib/value-format';
+
+import './cornelius.less';
+
+const momentInterval = {
+  daily: 'days',
+  weekly: 'weeks',
+  monthly: 'months',
+  yearly: 'years',
+};
+
+const timeLabelFormats = {
+  daily: 'MMMM D, YYYY',
+  weekly: '[Week of] MMM D, YYYY',
+  monthly: 'MMMM YYYY',
+  yearly: 'YYYY',
+};
+
+const defaultOptions = {
+  initialDate: null,
+  timeInterval: 'monthly',
+  drawEmptyCells: true,
+  rawNumberOnHover: true,
+  displayAbsoluteValues: false,
+  initialIntervalNumber: 1,
+  maxColumns: Infinity,
+
+  title: null,
+  timeColumnTitle: 'Time',
+  peopleColumnTitle: 'People',
+  stageColumnTitle: '{{ @ }}',
+  numberFormat: '0,0[.]00',
+  percentFormat: '0.00%',
+  timeLabelFormat: timeLabelFormats.monthly,
+
+  colors: {
+    min: '#ffffff',
+    max: '#041d66',
+    steps: 7,
+  },
+};
+
+function prepareOptions(options) {
+  options = extend({}, defaultOptions, options, {
+    initialDate: moment(options.initialDate),
+    colors: extend({}, defaultOptions.colors, options.colors),
+  });
+
+  return extend(options, {
+    timeLabelFormat: timeLabelFormats[options.timeInterval],
+    formatNumber: createNumberFormatter(options.numberFormat),
+    formatPercent: createNumberFormatter(options.percentFormat),
+    getColorForValue: chroma.scale([options.colors.min, options.colors.max])
+      .mode('hsl')
+      .domain([0, 100])
+      .classes(options.colors.steps),
```

Comment:
This is different from the original Cornelius - the color thresholds are divided **evenly** through the 7 steps while in the original it's:
```js
var repeatLevels = {
  'low': [0, 10],
  'medium-low': [10, 20],
  'medium': [20, 30],
  'medium-high': [30, 40],
  'high': [40, 60], // <- 20 diff
  'hot': [60, 70],
  'extra-hot': [70, 100] // <- 30 diff
}
```
This results in different color accentuation, which might not be an issue - I never worked with cohorts so I dunno if this matters.

Reply:
I think it's not an issue, especially considering that original Cornelius scale is very irregular (cannot choose right word for this) - it's not linear, nor logarithmic, not exponential - it's just hard-coded sequence which even does not always grow (10-10-10-10-**20**-10-**30**). I built a graph of that old pseudo-function and of linear and logarithmic functions on the same domain - linear was closest to original. It looks good enough, and visual changes are very subtle.

Maybe, later we could add a new setting to allow users to choose interpolation function from some list.

Reply:
I think it's not an issue, especially considering that original Cornelius scale is very irregular (cannot choose right word for this) - it's not linear, nor logarithmic, not exponential - it's just hard-coded sequence which even does not always grow (10-10-10-10-**20**-10-**30**). I built a graph of that old pseudo-function and of linear and logarithmic functions on the same domain - linear was closest to original. It looks good enough, and visual changes are very subtle.

Maybe, later we could add a new setting to allow users to choose interpolation function from some list.

Reply:
Whatever you see fit. I have no experience with such vizs.

(Btw maybe changing steps from 7 to 10 could be a good compromise)

Reply:
Let's keep 7 for now, later I planned add a setting for it :+1: